### DISCUSSION TITLE
CppAD::cg::CGException: what():  GreaterThanZero cannot be called for non-parameters

### DISCUSSION THREAD

**Discussion by IoannisDadiotis** [coin-or/CppAD#179] [CppADCodeGen]
Hi,

I am using CppAD in a robotics application for differentiating the robot forward kinematics. Sorry for not providing a compact piece of code to reproduce the issue for now, I will try to describe the problem. My overall application is using CppAD through a [CPPAdInterface](https://github.com/leggedrobotics/ocs2/blob/main/ocs2_core/include/ocs2_core/automatic_differentiation/CppAdInterface.h). The idea is that the user defines the function [adFunction_](https://github.com/leggedrobotics/ocs2/blob/ebde452b10d0eceaac45364f7bb8f0ac1038b637/ocs2_core/include/ocs2_core/automatic_differentiation/CppAdInterface.h#L226) to be differentiated. Then the models are created, compiled with CppAD and saved through [createModels](https://github.com/leggedrobotics/ocs2/blob/ebde452b10d0eceaac45364f7bb8f0ac1038b637/ocs2_core/src/automatic_differentation/CppAdInterface.cpp#L71).

In my case the `adFunction_` is defined as the orientation of a robot frame, which is given as rotation matrix from the software used for kinematics (`pinocchio` in my case) but has to be converted to euler angles through the [Eigen::MatrixBase< Derived >::eulerAngles](https://eigen.tuxfamily.org/dox/group__Geometry__Module.html#title20).
So my code looks like this:
```Cpp
ad_vector_t getOrientationCppAd(PinocchioInterfaceCppAd& pinocchioInterfaceCppAd,
                                                       const PinocchioStateInputMapping<ad_scalar_t>& mapping,
                                                       const ad_vector_t& state) {
  using ad_quaternion_t = Eigen::Quaternion<ad_scalar_t>;

  const auto& model = pinocchioInterfaceCppAd.getModel();
  auto& data = pinocchioInterfaceCppAd.getData();
  const ad_vector_t q = mapping.getPinocchioJointPosition(state);

  pinocchio::forwardKinematics(model, data, q);
  pinocchio::updateFramePlacements(model, data);

  ad_vector_t orientations(3 * endEffectorFrameIds_.size());
  for (int i = 0; i < endEffectorFrameIds_.size(); i++) {
    const size_t frameId = endEffectorFrameIds_[i];
    const auto euler = data.oMf[frameId].rotation().eulerAngles(0, 1, 2);
    orientations.segment<3>(3 * i) = euler;
  }
  return orientations;
}
```
, where `getOrientationCppAd` added as `adFunction_` of a `CppAdInterface` object.
The type declarations used can be found [here](https://github.com/leggedrobotics/ocs2/blob/main/ocs2_core/include/ocs2_core/automatic_differentiation/Types.h).

I have noticed that when using the `eulerAngles()` function I get the following error during compiling with CppAD.
```
terminate called after throwing an instance of 'CppAD::cg::CGException'
what():  GreaterThanZero cannot be called for non-parameters
```

More in particular I have been trying to trace the problem inside `eulerAngles` function and it seems that the problem was found in this [line](https://gitlab.com/libeigen/eigen/-/blob/73922b0174f84e78f486eb89de86dffc881acbcb/Eigen/src/Geometry/EulerAngles.h#L91) when evaluating `(!odd)` inside the if statement. However this expression inside an if condition is also used in another [line](https://gitlab.com/libeigen/eigen/-/blob/73922b0174f84e78f486eb89de86dffc881acbcb/Eigen/src/Geometry/EulerAngles.h#L56) above which does not cause a problem during compilation.

Do you have an idea of what may be the issue?

**Comment by bradbell** [comment]
Are you able to run the same code without using CppADCodeGen; i.e, just using CppAD ?

**Comment by IoannisDadiotis** [comment]
Since I am quite new to CppAD and CppADCodeGen and I am only using it inside an existing open source [library](https://leggedrobotics.github.io/ocs2/from_urdf_to_ocp.html?highlight=differentiation) I am not sure how can I do this.

**Comment by IoannisDadiotis** [comment]
Just a guess for what can be a solution, based on [previous issues](https://github.com/joaoleal/CppADCodeGen/issues/56) I may have to replace the condition below
`if((odd && res[0]<SCALAR_T(0)) || ((!odd) && res[0]>SCALAR_T(0))) `
with a combination of `CppAD::CondExpOp`

**Comment by bradbell** [comment]
Yes, that should solve your problem, but it involves changing eigen's code ?

**Comment by IoannisDadiotis** [comment]
indeed, I have managed to work around by re implementing the eigen function I needed using `CppAD::CondExpOp`.

Please let me know if you are aware of a different way without modifying eigen's source code, otherwise the issue can be closed. thanks
